File: doc-examples/example-kotlin/src/main/kotlin/example/ProfileService.kt
package example

import io.micronaut.objectstorage.ObjectStorageEntry
import io.micronaut.objectstorage.ObjectStorageOperations
import io.micronaut.objectstorage.request.ListObjectsRequest
import io.micronaut.objectstorage.request.UploadRequest
import io.micronaut.objectstorage.response.ListObjectsResponse
import jakarta.inject.Singleton
import java.io.File
import java.io.InputStream
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardCopyOption

//tag::beginclass[]
@Singleton
open class ProfileService(private val objectStorage: ObjectStorageOperations<*, *, *>) {
//end::beginclass[]

    //tag::upload[]
    open fun saveProfilePicture(userId: String, path: Path): String? {
        val request = UploadRequest.fromPath(path, userId) // <1>
        val response = objectStorage.upload(request) // <2>
        return response.key // <3>
    }
    //end::upload[]

    //tag::input-stream-upload[]
    open fun saveProfilePicture(userId: String, inputStream: InputStream, contentLength: Long): String? {
        val request = UploadRequest.fromInputStream(
            inputStream,
            "$userId/profile.jpg",
            "image/jpeg",
            contentLength // <1>
        )
        request.metadata = mapOf("source" to "validated-upload") // <2>
        return objectStorage.upload(request).key
    }
    //end::input-stream-upload[]

    //tag::retrieve[]
    open fun retrieveProfilePicture(userId: String, fileName: String): Path? {
        val key = "$userId/$fileName"
        val stream = objectStorage.retrieve<ObjectStorageEntry<*>>(key) // <1>
            .map { obj: ObjectStorageEntry<*> -> obj.inputStream }

        return if (stream.isPresent) {
            val destination = File.createTempFile(userId, "temp").toPath()
            Files.copy(stream.get(), destination, StandardCopyOption.REPLACE_EXISTING)
            destination
        } else {
            null
        }
    }
    //end::retrieve[]

    //tag::delete[]
    open fun deleteProfilePicture(userId: String, fileName: String) {
        val key = "$userId/$fileName"
        objectStorage.delete(key) // <1>
    }
    //end::delete[]

    //tag::paginated-listing[]
    open fun listProfilePicturesByPage(userId: String) {
        val prefix = "$userId/" // raw prefix filtering
        var continuation: String? = null
        do {
            val request = ListObjectsRequest(2, prefix, continuation) // <1>
            val response: ListObjectsResponse = objectStorage.listObjects(request)

            // Process the keys in this page
            response.keys.forEach { key -> println("Found: $key") } // <2>

            // Replay the continuation token returned by the provider for the next page
            continuation = response.continuationToken.orElse(null) // <3>
        } while (continuation != null)
    }
    //end::paginated-listing[]

//tag::endclass[]
}
//end::endclass[]

File: doc-examples/example-python/src/main/python/example/ProfileService.py
import logging

from jakarta.inject import Singleton
from java.io import File, IOException, InputStream
from java.nio.file import Files, Path, StandardCopyOption
from micronaut.objectstorage import ObjectStorageOperations
from micronaut.objectstorage.request import ListObjectsRequest, UploadRequest

LOG = logging.getLogger(__name__)


# tag::beginclass[]
@Singleton
class ProfileService:

    def __init__(self, object_storage: ObjectStorageOperations):
        self.object_storage = object_storage
# end::beginclass[]

    # tag::upload[]
    def save_profile_picture(self, user_id: str, path: Path) -> str:
        request = UploadRequest.fromPath(path, user_id)  # <1>
        response = self.object_storage.upload(request)  # <2>
        return response.getKey()  # <3>
    # end::upload[]

    # tag::input-stream-upload[]
    def save_profile_picture_stream(self, user_id: str, input_stream: InputStream, content_length: int) -> str:
        request = UploadRequest.fromInputStream(
            input_stream,
            user_id + "/profile.jpg",
            "image/jpeg",
            content_length  # <1>
        )
        request.setMetadata({"source": "validated-upload"})  # <2>
        return self.object_storage.upload(request).getKey()
    # end::input-stream-upload[]

    # tag::retrieve[]
    def retrieve_profile_picture(self, user_id: str, file_name: str) -> Path | None:
        destination = None
        try:
            key = user_id + "/" + file_name
            stream = self.object_storage.retrieve(key).map(lambda entry: entry.getInputStream())  # <1>
            if stream.isPresent():
                destination = File.createTempFile(user_id, "temp").toPath()
                Files.copy(stream.get(), destination, StandardCopyOption.REPLACE_EXISTING)
                return destination
            else:
                return None
        except IOException as e:
            LOG.error("Error while trying to save profile picture to the local file [%s]: %s", destination, e.getMessage())
            return None
    # end::retrieve[]

    # tag::delete[]
    def delete_profile_picture(self, user_id: str, file_name: str) -> None:
        key = user_id + "/" + file_name
        self.object_storage.delete(key)  # <1>
    # end::delete[]

    # tag::paginated-listing[]
    def list_profile_pictures_by_page(self, user_id: str) -> None:
        prefix = user_id + "/"  # raw prefix filtering
        continuation = None
        while True:
            request = ListObjectsRequest(2, prefix, continuation)  # <1>
            response = self.object_storage.listObjects(request)

            # Process the keys in this page
            for key in response.getKeys():
                print("Found: " + key)  # <2>

            # Replay the continuation token returned by the provider for the next page
            continuation = response.getContinuationToken().orElse(None)  # <3>
            if continuation is None:
                break
    # end::paginated-listing[]

# tag::endclass[]
# end::endclass[]
